### PR TITLE
Allow to pass metadata directly in create and update mutations for product app models

### PR DESCRIPTION
**Pull request by SzymJ**
I want to merge this change because it adds possibility to pass `metadata` and `privateMetadata` directly in `create/update` mutations for product app models.

 

# Impact

* [ ] New migrations
* [x] New/Updated API fields or mutations
* [ ] Deprecated API fields or mutations
* [ ] Removed API types, fields, or mutations
* [ ] Documentation needs to be updated

# Pull Request Checklist

* [ ] Privileged queries and mutations are guarded by proper permission checks
* [ ] Database queries are optimized and the number of queries is constant
* [ ] Database migration files are up to date
* [ ] The changes are tested
* [ ] GraphQL schema and type definitions are up to date
* [ ] Changes are mentioned in the changelog

### CODE REVIEW

**Comment by patrys** [comment]
Task linked: [SALEOR-8010 Allow metadata to be passed to Category, Collection, Product, ProductVariant, DigitalContent create mutations](https://app.clickup.com/t/2549495/SALEOR-8010)

File: saleor/graphql/core/mutations.py [hunk 1]
```diff
@@ -575,13 +575,30 @@ def perform_mutation(cls, _root, info, **data):
         instance = cls.get_instance(info, **data)
         data = data.get("input")
         cleaned_input = cls.clean_input(info, instance, data)
+        metadata_list = cleaned_input.pop("metadata", None)
```
Comment (korycins) [inline saleor/graphql/core/mutations.py:578]: Maybe it is already covered but shouldn't we validate the permission of writing to metadata here?

File: saleor/graphql/product/mutations/products.py [hunk 2]
```diff
@@ -78,9 +80,21 @@ class CategoryInput(graphene.InputObjectType):
...
+        MetadataInput,
+        description=(
+            "Fields required to update the category private metadata." + ADDED_IN_38
+        ),
+        required=False,
+    )
 
 
-class CategoryCreate(ModelMutation):
+class CategoryCreate(MutationWithMetadataMixin, ModelMutation):
```
Comment (korycins) [inline saleor/graphql/product/mutations/products.py:97]: I am not convinced about the current approach. Some metadata logic is placed in `MutationWithMetadataMixin` and some logic is placed `ModelMutation`.  It is hard to get full context jumping between mixin and model mutation. 
Can't we extend `ModelMutation` to have full metadata support?
The support should cover, validating input, validating permission for given instance, saving in the instance etc. 
We could control it by setting the flag in Meta input. Something like:
```
class CategoryCreate(ModelMutation):
    class Arguments:
         ...
         <some input with metadata>
    class Meta:
        ...
        support_meta_field=True,
        support_private_meta_field=True
```
Based on provided flags (i didn't think a lot about the naming of the fields, just use the first one to show the proposed solution), ModelMutation can control if should handle metadata fields from the input or not. Thanks to this, as a developer, one thing that you need to remember during adding new meta fields is to provide correct input fields and setting the flags to correct value.
I think that we could provide similar flow to BaseMutation but more generic (Probably we could add it only to BaseMutation, and use it in ModelMutation)

@SzymJ @maarcingebala  What do you think?

Comment (IKarbowiak) [reply saleor/graphql/product/mutations/products.py:97]: cc. @fowczarek as he's the author of the current metadata solution :)

Comment (fowczarek) [reply saleor/graphql/product/mutations/products.py:97]: IMO solution proposed by @korycins is better.

Comment (SzymJ) [reply saleor/graphql/product/mutations/products.py]: Done, I moved all methods to `BaseMutation`

**Review by IKarbowiak — APPROVED** [review]

**Review by korycins — APPROVED** [review]

**Review by maarcingebala — APPROVED** [review]

**Pull request merged**